hadolint/hadolint
# DL3061: Check for instruction order

Not all dockerfiles are valid just because all instructions in the
dockerfile are individually valid. For example, a `RUN` instruction may
not precede the first `FROM` instruction and so on. The new rule DL3061
checks that the order of the instructions is not completely bogus. That
means the Dockefile must begin with a comment, an `ARG` or a `FROM`
instruction.

- Add new rule DL3061 to check for instruction order
- Add tests for rule DL3061
- Fix regression tests wrt. rule DL3061

fixes: #737

### How to verify it
The new rule should alert in case the beginning of a dockerfile messes up instruction order such that a build can not succeed, e.g.:
```Dockerfile
RUN foo bar  # this is bad
FROM alpine:3
RUN bar foo
```
But should of course not alert in case a Dockerfile is valid and can be build:
```Dockerfile
# comments are ok
ARG BUILD_ARGUMENTS_OK=true
FROM alpine:3  # FROM must precede other instructions
LABEL foo "bar" # this is ok now
RUN something  # other instructions can follow
```

Thanks to @ye for suggesting this rule.
